src/test/java/com/google/sps/SentimentServletTest.java

```diff
@@ -50,100 +57,156 @@ public void setUp() throws IOException {
-    when(requestMock.getReader()).thenReturn(new BufferedReader(new StringReader(json)));
+  public void noCommentsOrCaptionsSendsHttpResponseError() throws IOException {
+    List<String> comments = Arrays.asList();
+    String captions = "";
+    
+    when(requestMock.getParameter("video-id")).thenReturn(VIDEO_ID);
+    when(commentServiceMock.getCommentsFromId(VIDEO_ID)).thenReturn(comments);
+    when(captionServiceMock.getCaptionFromId(VIDEO_ID)).thenReturn(captions);
+
+    sentimentServlet.doGet(requestMock, responseSpy);
     
+    String expectedNoData = "{\"dataAvailable\":false}\n";
+
+    verify(responseSpy).sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR,
+        "No comments or captions available to analyze.");
+    Assert.assertEquals(expectedNoData, stringWriter.toString());
+  }
+
+  /**
+   * When CommentService throws an IllegalArgumentException, a 500 error should be sent.
+   */
+  @Test
+  public void invalidVideoIdSendsHttpResponseError() throws IOException {
+    when(requestMock.getParameter("video-id")).thenReturn(VIDEO_ID);
+    when(commentServiceMock.getCommentsFromId(VIDEO_ID))
+        .thenThrow(IllegalArgumentException.class);
+
     sentimentServlet.doGet(requestMock, responseSpy);
 
     verify(responseSpy).sendError(HttpServletResponse.SC_BAD_REQUEST,
-        "No comments to analyze. Request must be a non-empty array.");
+        "Video is private or does not exist.");
   }
 
   /**
-   * If the Natural Language API fails and throws an ApiException, a 500 error is sent.
+   * CommentService throws an GoogleJsonResponseException, a 500 error should be sent and only the
+   * captions are analyzed.
    */
   @Test
-  public void languageServiceFailureSendsHttpResponseError() throws IOException {
-    when(requestMock.getReader()).thenReturn(new BufferedReader(new StringReader(VALID_COMMENTS)));
-    when(languageServiceMock.analyzeSentiment(any(Document.class))).thenThrow(ApiException.class);
+  public void commentRetrievalFailureSendsHttpResponseError() throws IOException {
+    String captions = "foo. bar. foobar.";
+
+    when(requestMock.getParameter("video-id")).thenReturn(VIDEO_ID);
+    when(commentServiceMock.getCommentsFromId(VIDEO_ID))
+        .thenThrow(GoogleJsonResponseException.class);
+    when(captionServiceMock.getCaptionFromId(VIDEO_ID)).thenReturn(captions);
+    when(languageServiceMock.analyzeSentiment(any(Document.class)).getDocumentSentiment())
+        .thenReturn(sentimentMock);
+    when(sentimentMock.getScore()).thenReturn(SCORE);
+    doNothing().when(languageServiceMock).close();
 
     sentimentServlet.doGet(requestMock, responseSpy);
 
+    String expected = "{\"score\":" + df.format(SCORE) + ",\"dataAvailable\":true}\n";
+
     verify(responseSpy).sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR,
-        "Language service client failed.");
+        "Comments could not be retrieved.");
+    Assert.assertEquals(expected, stringWriter.toString());
```

Comment:
Is it okay to send both the error and the value to the front end? For example, if I want to get the score in the front end, do I find it in fetch..then.. or fetch..catch.. ?

Reply:
nvm now I know you are reading the status and the message